helix-vcs/src/git.rs

```diff
@@ -103,3 +340,128 @@ fn find_file_in_commit(repo: &Repository, commit: &Commit, file: &Path) -> Optio
+                        //
+                        // With the current implementation, there's no way we can handle huge files
+                        // that do not fit into memory. Maybe we can set a size limit? Anything
+                        // over a certain size will simply be read twice: once for getting the
+                        // normalized size, and once for the hasher updates?
+                        const BUFFER_SIZE: usize = 8 * 1024;
+                        let mut buffer = [0u8; BUFFER_SIZE];
+
+                        let mut len = file.read(&mut buffer)?;
+                        if content_inspector::inspect(&buffer[..len])
+                            == content_inspector::ContentType::BINARY
+                        {
+                            // No CRLF handling! We update the part already read + the remaining
+                            // content in the file.
+                            hasher.update(format!("{}", meta.len()).as_bytes());
+                            hasher.update(b"\0");
+
+                            hasher.update(&buffer[..len]);
+                            std::io::copy(&mut file, &mut hasher)?;
+                        } else {
+                            // It's a text file. CRLF transformation as planned.
+                            let mut normalized_file = Vec::with_capacity(meta.len() as usize);
+                            let mut was_cr = false;
+
+                            loop {
+                                buffer[..len].iter().for_each(|byte| {
+                                    if was_cr && *byte == b'\n' {
+                                        normalized_file.pop();
+                                    }
+                                    normalized_file.push(*byte);
+                                    was_cr = *byte == b'\r';
+                                });
+
+                                if len < BUFFER_SIZE {
+                                    break;
+                                }
+                                len = file.read(&mut buffer)?;
+                            }
+
+                            hasher.update(format!("{}", normalized_file.len()).as_bytes());
+                            hasher.update(b"\0");
+
+                            hasher.update(&normalized_file);
+                        }
+                    } else {
+                        hasher.update(format!("{}", meta.len()).as_bytes());
+                        hasher.update(b"\0");
+
+                        std::io::copy(&mut file, &mut hasher)?;
+                    }
+
+                    let hash: [u8; 20] = hasher.finalize().into();
+                    ObjectId::from(hash)
+                };
+
+                Some((entry_mode, oid))
+            } else {
+                // It's a non-symlink folder. Git doesn't track folders. Same as deletion.
+                None
+            }
+        }
+        Err(_) => None,
+    })
```

Comment:
the indentation levels here can be reduced by using this pattern, it would avoid the line `Err(_) => None` at the end that is difficult to link to the earlier `path.metadata()` call and the `if meta.is_file()` that is 99% of the `Ok` branch of the match

```rust
let metadata = match path.metadata() {
    Ok(m) if m.is_file() => m,
    // Ignore errors and non-files
    _ => return Ok(None),
}

/// ...
```